### PR TITLE
Fix #102, clean up sc_loads_test

### LINKED ISSUE #102: Stack smashing in AtsEntryOverflow test
**Checklist (Please check before submitting)**

* [x] I reviewed the [Contributing Guide](https://github.com/nasa/SC/blob/main/CONTRIBUTING.md).
* [x] I performed a cursory search to see if the bug report is relevant, not redundant, nor in conflict with other tickets.

**Describe the bug**
It appears this test case takes things a bit too literally - it creates an ATS buffer on the stack, then writes past the end of it.

**To Reproduce**
Execute the SC "loads" coverage test with stack protection enabled.  This will abort due to stack corruption.

**Expected behavior**
Should execute successfully.

**System observed on:**
Debian

**Additional context**
The test case cannot literally overrun the buffer.  It should just initialize the ATS so it looks like it will overrun the buffer - that is, set the length field too large - without writing those bytes.

**Reporter Info**
Joseph Hickey, Vantage Systems, Inc.

### PR DESCRIPTION
**Pull request by jphickey**
**Checklist (Please check before submitting)**

* [x] I reviewed the [Contributing Guide](https://github.com/nasa/SC/blob/main/CONTRIBUTING.md).
* [x] I signed and emailed the appropriate [Contributor License Agreement](https://github.com/nasa/cFS/blob/main/CONTRIBUTING.md#contributor-license-agreement-cla) to GSFC-SoftwareRelease@mail.nasa.gov and copied cfs-program@lists.nasa.gov.

**Describe the contribution**
Significant buffer management issues existed in many/all of these test cases and was causing stack corruption.  Rewriting the test cases to better follow current recommended practices, and reduce repetition.

Fixes #102

**Testing performed**
Build and run sc_loads_test, confirm coverage is maintained

**Expected behavior changes**
No more stack overflow/smashing.

**System(s) tested on**
Debian

**Additional context**
Test case code reduced by about 40%, should be much easier to maintain going forward.

**Contributor Info - All information REQUIRED for consideration of pull request**
Joseph Hickey, Vantage Systems, Inc.

### CODE REVIEW

**Comment by skliper** [comment]
@jphickey - I already fixed all of this with https://github.com/nasa/SC/pull/98.  Whatever this commit is on doesn't seem in sync w/ main.

See my comment: https://github.com/nasa/SC/issues/102#issuecomment-1715833242

**Review by skliper — DISMISSED** [review]
This looks like a backwards step from #98 since it reintroduces local tables instead of initializing them in setup.

**Comment by jphickey** [comment]
> @jphickey - I already fixed all of this with #98. Whatever this commit is on doesn't seem in sync w/ main.
> 
> See my comment: [#102 (comment)](https://github.com/nasa/SC/issues/102#issuecomment-1715833242)

Yep, I noticed that!  You had already fixed the bulk of what I had noticed when trying to run this test - buffer overruns galore.  I rebased my change.  

My change is still worthwhile to consolidate a lot of the repeated logic and use more of the modern test macros - e.g. I also got rid of all remaining "UtAssert_True" checks in favor of the more targeted version.

**Comment by jphickey** [comment]
I see that you instantiated the UT tables in `unit-test/utilities/sc_test_utils.c` ... I will update this to use those tables rather than an instance in the loads test file.

**Comment by skliper** [comment]
Your change now deviates from the pattern I employed for the rest of the SC tests, with the tables defined and initialized in the setup.  I hadn't fixed the UtAssert_Trues which would be a good fix but I'd suggest sticking with the pattern established in #98 for the tables.  After that, I had also consolidated some logic to avoid repeats... really hard to tell based on how this comment was done what's better vs just different since your change wasn't based off main to begin with.  I'm not sure a step backwards from consistency employed for the other tests is helpful.

**Comment by jphickey** [comment]
@skliper - not intending to take a "step backwards" -- I like the pattern there in #98 and will use it, its just that I was basing this off `equuleus-rc1` tag that didn't have that rework yet (because this tag is what the stakeholder is using).  Which is unfortunate - https://github.com/nasa/SC/pull/98#issuecomment-1721234040

I will update this to use the shared table instances.  That will cut down on repetition even more.  I just wish we had merged that when you first put in the PR.  Would have saved some duplicate work.

**Comment by jphickey** [comment]
@skliper -- What's left in here is basically consolidation of repeated code during init of the ATS/RTS tables.  Should make the test cases easier to maintain going forward.

File: unit-test/sc_loads_tests.c [hunk 1]
```diff
@@ -40,368 +40,404 @@
...
-    MsgSize2 = (SC_ATS_BUFF_SIZE32 - SC_ATS_HDR_WORDS + 4 - j) * SC_BYTES_IN_WORD;
-    UT_SetDataBuffer(UT_KEY(CFE_MSG_GetSize), &MsgSize2, sizeof(MsgSize2), false);
+    uint8 AtsIndex = 0;
 
-    /* Set up final entry that will create condition */
-    j += ((MsgSize2 + SC_ROUND_UP_BYTES) / SC_BYTES_IN_WORD) + SC_ATS_HDR_WORDS;
-    Entry            = (SC_AtsEntryHeader_t *)&SC_OperData.AtsTblAddr[AtsIndex][j];
-    Entry->CmdNumber = i + 1;
-    UT_SetDataBuffer(UT_KEY(CFE_MSG_GetSize), &MsgSize1, sizeof(MsgSize1), false);
+    UT_SC_GetAtsTable(AtsIndex);
```
Comment (skliper) [inline unit-test/sc_loads_tests.c:400]: What are the calls to UT_SC_GetAtsTable for when you don't use the returned value?  Looks like a repeated pattern in a handful of tests.

**Comment by skliper** [comment]
@jphickey - great to see the UtAssert_True's go away also!  Maybe that API could be deprecated some day.

**Comment by dzbaker** [comment]
CCB 21 September 2023: Provisionally approved pending re-review by @skliper.

**Comment by skliper** [comment]
@dzbaker - review is pending response to my previous comment

File: unit-test/sc_loads_tests.c (hunk 1 above)
Comment (jphickey) [reply unit-test/sc_loads_tests.c:400]: Thanks for catching that, you are correct this has no effect after the rebase.  Removing these calls.

**Comment by jphickey** [comment]
Updated based on @skliper review comment and rebased to resolve merge conflict.  Should be good to go now.

**Review by skliper — APPROVED** [review]
Great set of updates!

**Pull request merged**